### PR TITLE
feat: Add namespace filtering to tolgee pull #17

### PR DESCRIPTION
**Pull request by shixin-guo**

### CODE REVIEW

**Comment by shixin-guo** [comment]
@cyyynthia Hi, I have the same requirement  as [#17], so I support it myself firstly, if possible pls take a look, and help review it

File: package.json [hunk 1]
```diff
@@ -17,7 +17,6 @@
     "pretest:e2e": "npm run build && npm run tolgee:start",
     "posttest:e2e": "npm run tolgee:stop",
     "test:e2e": "cross-env NODE_OPTIONS=\"--experimental-vm-modules\" jest -c jest.e2e.config.ts --runInBand",
-    "test:e2e-run": "cross-env NODE_OPTIONS=\"--experimental-vm-modules\" jest -c jest.e2e.config.ts --runInBand",
```
Comment (shixin-guo) [inline package.json:20]: same as "test:e2e", so I remove it

Comment (cyyynthia) [reply package.json:20]: It is necessary. It's not the same because it does not have pre/post tasks, and CI uses it (so it sets up the docker container itself).

Comment (shixin-guo) [reply package.json]: Got, fixed

File: src/commands/pull.ts [hunk 2]
```diff
@@ -45,6 +46,7 @@ export default new Command()
     '<path>',
     'Destination path where translation files will be stored in'
   )
+
```
Comment (cyyynthia) [inline src/commands/pull.ts:49]: Whitespace

File: src/options.ts [hunk 3]
```diff
@@ -34,6 +34,7 @@ export type BaseOptions = {
   apiKey: string;
   projectId: number;
   client: Client;
+  namespace?: string[];
```
Comment (cyyynthia) [inline src/options.ts:37]: You shouldn't add it to the global object. Add it to the the type specific to pull.ts

File: src/commands/pull.ts [hunk 4]
```diff
@@ -71,6 +73,12 @@ export default new Command()
       .default('.')
       .argParser((v) => v || '')
   )
+  .addOption(
+    new Option(
+      '-n, --namespace <namespace...>',
+      'Namespace, pull keys from specified namespaces. Defaults all namespaces'
```
Comment (cyyynthia) [inline src/commands/pull.ts:79]:
```suggestion
      '-n, --namespaces <namespaces...>',
      'List of namespaces to pull. Defaults to all namespaces'
```

File: src/commands/pull.ts [hunk 5]
```diff
@@ -71,6 +73,12 @@ export default new Command()
...
+    ).argParser((v, a: string[]) => [v, ...(a || [])])
```
Comment (cyyynthia) [inline src/commands/pull.ts:80]: This parser isn't needed here. It was used in the states argument to turn the argument always uppercase as expected by the API; but for example the languages does not use this.

**Review by cyyynthia — CHANGES_REQUESTED** [review]
Some minor nitpicks, otherwise great job! Huge thanks for adding a test for this feature as well ❤️

**Comment by cyyynthia** [comment]
In addition, I also worry about the error handling, if the user doesn't specify a valid namespace. Can you double check this, and make sure the error is correctly handled, with proper error reporting to the user?

**Comment by shixin-guo** [comment]
> In addition, I also worry about the error handling, if the user doesn't specify a valid namespace. Can you double check this, and make sure the error is correctly handled, with proper error reporting to the user?

@cyyynthia Thank you for your suggestion. 😄 I have addressed most of the issues, and regarding this particular requirement, I believe it's a good idea. However, I've encountered a problem with our tolgee-platform. It returns an HTTP 400 error when the Namespace is valid, and this error gets caught by your global error handler. To enhance support for this feature, it would be beneficial to introduce a specific error code that can inform the CLI about the error. Currently, the only way to verify the namespace is by making an API call, but I think this approach might be overly complex.

**Comment by shixin-guo** [comment]
By the way, I've noticed that when we use Commander.js, the .**choices** method (part of the addOption function) should come after **argParser**. Otherwise, choices won't function as expected. However, I believe this issue isn't directly related to the current PR. Do we need to address this?
image

**Comment by cyyynthia** [comment]
> It returns an HTTP 400 error when the Namespace is valid, and this error gets caught by your global error handler. To enhance support for this feature, it would be beneficial to introduce a specific error code that can inform the CLI about the error. Currently, the only way to verify the namespace is by making an API call, but I think this approach might be overly complex.

`client.export.export` call in pull.ts will throw an error on non-success HTTP response codes. By wrapping it in a try/catch, you can capture the error and perform some validation on it.

```js
try {
  // HTTP call
} catch (e) {
  if (e instanceof HttpError && e.response.status === 400) {
    // Handle error...
    return;
  }

  throw e;
}
```

Since `e.response` will be the raw response object, you can call `.json()` on it to fetch the error object with the error code. If checking the error code, make sure to return only in the branch where the error is handled, so the error can be propagated to the global handler if necessary.

Doing this might in specific cases cause verbose debugging to have issues (since it will also try to consume the body). A workaround could be to implement `.json()` and `.text()` on HttpError, and have them retain the response body text in its own properties (and only call into `response.text()` the first time in `.text()`, and have `.json()` simply call `.text()` and `JSON.parse` it).

> By the way, I've noticed that when we use Commander.js, the .choices method (part of the addOption function) should come after argParser. Otherwise, choices won't function as expected.

After checking Commander.js source, it seems they use the argParser to implement the choices method. This means it should be part of the argParser the cli uses (since the idea was to let the user input them in lowercase and don't complain about it). https://github.com/tj/commander.js/blob/master/lib/option.js#L184-L186

It's unrelated but it can be tucked with this PR I think.

**Comment by shixin-guo** [comment]
> https://github.com/tj/commander.js/blob/master/lib/option.js#L184-L186

sure, I know, I used the same way as you suggested (try catch wrapper export function) yesterday,  I thought the current response was not clear and could not get the exact reasons, so I confirmed with you here ...whatever I can add this logic now !!!

**Comment by shixin-guo** [comment]
hi @cyyynthia ,  error handler added


and by the way, I found https://github.com/tj/commander.js/blob/4ef19faac1564743d8c7e3ce89ef8d190e1551b4/lib/option.js#L133C3-L136C4

if we use 'argParser' function,  it will override the 'parseArg' as well, so even if we change the order it also can not be perfect, I thought we needed to wait for commander.js to fix it,  (in my options, use a stack to record all argParser and run them one by one maybe good)

**Comment by cyyynthia** [comment]
> if we use 'argParser' function, it will override the 'parseArg' as well, so even if we change the order it also can not be perfect, I thought we needed to wait for commander.js to fix it, (in my options, use a stack to record all argParser and run them one by one maybe good)

Yeah that's what I meant; we need to implement our own validation in the argParser instead of relying on the builtin choices method. I do agree commanderjs should allow both to coexist but in the meantime we'll have to deal with it, haha. The good news is that currently it does validation but forces the value to be uppercase, which is not a big deal. It can be left as-is for now, unless you want to fix it.

File: src/commands/pull.ts [hunk 6]
```diff
@@ -21,21 +23,29 @@ async function fetchZipBlob(opts: PullOptions): Promise<Blob> {
...
+    const zipBlob = await loading(
+      'Fetching strings from Tolgee...',
+      fetchZipBlob(opts)
+    );
+    await loading('Extracting strings...', unzipBuffer(zipBlob, path));
+    success('Done!');
+  } catch (e) {
+    if (e instanceof HttpError && e.response.status === 400) {
+      const res = await e.response.text();
+      error(res);
```
Comment (cyyynthia) [inline src/commands/pull.ts:44]: This might be a little bit too brutal to dump the JSON error object to the user 😅

It'd be more friendly to show an appropriate error message in plain English explaining what's wrong.

**Review by cyyynthia — CHANGES_REQUESTED** [review]

File: src/commands/pull.ts (hunk 6 above)
Comment (shixin-guo) [reply src/commands/pull.ts]: Thank you for your patient and detailed guidance, haha

Comment (cyyynthia) [reply src/commands/pull.ts]: Super sorry for the late reply! Life is quite busy on my end 😅 

I think it'd be better to only return the error code instead of the raw json object, making it a tad more readable.

Comment (shixin-guo) [reply src/commands/pull.ts]: @cyyynthia  Can I use this function "e.getErrorText()"
```
  if (e instanceof HttpError) {
      error(
        `Please check if your parameters, including namespaces, are configured correctly. Tolgee responded with error: ${e.getErrorText()}`
      );
      return;
    }
```

Comment (cyyynthia) [reply src/commands/pull.ts]: This will not give the desired result; what you want is to get the body by using `await e.response.json()` and print the `obj.error` (I think those are the field names; I might be wrong but it's something along these lines. 🙂

Comment (shixin-guo) [reply src/commands/pull.ts]: @cyyynthia  thanks for your suggestion, I have pushed a commit to optimize this code, this is the screenshot:
Screenshot2023_10_16_103657

can you take a look when you are available

Comment (cyyynthia) [reply src/commands/pull.ts]: This seems to be the expected error code, since the actual error reported by Tolgee is an empty export 👍🏻

**Review by cyyynthia — APPROVED** [review]

**Comment by cyyynthia** [comment]
The E2E test failure is unrelated to your changes; merging as-is and I'll fix the issue (and release the flag you've added) today or tomorrow. Thanks for the PR!

**Comment by EtienneGuilletPackative** [comment]
@cyyynthia I think the doc is outdated on that part 
[https://tolgee.io/tolgee-cli/push-pull-strings](url)
image
